![Moneyball](https://proxy.duckduckgo.com/iu/?u=http%3A%2F%2F2.bp.blogspot.com%2F-L8BZ-HVVizs%2FToIogdMGV-I%2FAAAAAAAAAsk%2FeVODynuBxZE%2Fs1600%2Fmoneyball.png&f=1&nofb=1)

# Moneyball homework

## Rules of Baseball

You don't need to know much about Baseball to complete this exercise. If you're totally unfamiliar with Baseball, check out this useful [explanatory video](https://www.youtube.com/watch?v=0bKkGeROiPA)!

### [Background](https://www.youtube.com/watch?v=yGf6LNWY9AI)
*Source: Wikipedia*

### The 2002 Oakland A's

The Oakland Athletics' 2002 season was the team's 35th in Oakland, California. It was also the 102nd season in franchise history. The Athletics finished first in the American League West with a record of 103-59.

The Athletics' 2002 campaign ranks among the most famous in franchise history. Following the 2001 season, Oakland saw the departure of three key players (the lost boys). Billy Beane, the team's general manager, responded with a series of under-the-radar free agent signings. The new-look Athletics, despite a comparative lack of star power, surprised the baseball world by besting the 2001 team's regular season record. The team is most famous, however, for winning 20 consecutive games between August 13 and September 4, 2002.[1] The Athletics' season was the subject of Michael Lewis' 2003 book Moneyball: The Art of Winning an Unfair Game (as Lewis was given the opportunity to follow the team around throughout that season)

This project is based off the book written by Michael Lewis (later turned into a movie).
### Moneyball Book

The central premise of book Moneyball is that the collective wisdom of baseball insiders (including players, managers, coaches, scouts, and the front office) over the past century is subjective and often flawed. Statistics such as stolen bases, runs batted in, and batting average, typically used to gauge players, are relics of a 19th-century view of the game and the statistics available at that time. The book argues that the Oakland A's' front office took advantage of more analytical gauges of player performance to field a team that could better compete against richer competitors in Major League Baseball (MLB).

Rigorous statistical analysis had demonstrated that on-base percentage and slugging percentage are better indicators of offensive success, and the A's became convinced that these qualities were cheaper to obtain on the open market than more historically valued qualities such as speed and contact. These observations often flew in the face of conventional baseball wisdom and the beliefs of many baseball scouts and executives.

By re-evaluating the strategies that produce wins on the field, the 2002 Athletics, with approximately US 44 million dollars in salary, were competitive with larger market teams such as the New York Yankees, who spent over US$125 million in payroll that same season.

Because of the team's smaller revenues, Oakland is forced to find players undervalued by the market, and their system for finding value in undervalued players has proven itself thus far. This approach brought the A's to the playoffs in 2002 and 2003.

In this project we'll work with some data and with the goal of trying to find replacement players for the ones lost at the start of the off-season - During the 2001–02 offseason, the team lost three key free agents to larger market teams: 2000 AL MVP Jason Giambi to the New York Yankees, outfielder Johnny Damon to the Boston Red Sox, and closer Jason Isringhausen to the St. Louis Cardinals.

The main goal of this project is for you to feel comfortable working with Python on real data to try and derive actionable insights!
Let's get started!

***Follow the steps outlined in bold below using your new Python skills and help the Oakland A's recruit under-valued players!***

![Gameplan](https://proxy.duckduckgo.com/iu/?u=https%3A%2F%2Ftse1.mm.bing.net%2Fth%3Fid%3DOIP.RI7hlTMbnkPlzADQTtMeqwHaE8%26pid%3DApi&f=1)

### Data

We'll be using data from Sean Lahaman's Website a very useful source for baseball statistics. The documentation for the csv files is located in the readme2013.txt file. You may need to reference this to understand what acronyms stand for.

**Use Pandas to open the Batting file and assign it to a dataframe called batting**

In [1]:
# Define the url the data is located at
batting_url = 'https://rotterdamai001.blob.core.windows.net/python/mlb/core/Batting.csv'

In [2]:
# Import pandas for data processing
import pandas as pd
# Save batting dataframe
batting = pd.read_csv(batting_url)

**Review the first 5 rows of the batting file**

In [3]:
# Output the first five rows in the console
batting.head()

playerID  yearID  stint teamID lgID   G   AB   R   H  2B  ...   RBI   SB  \
0  abercda01    1871      1    TRO  NaN   1    4   0   0   0  ...   0.0  0.0   
1   addybo01    1871      1    RC1  NaN  25  118  30  32   6  ...  13.0  8.0   
2  allisar01    1871      1    CL1  NaN  29  137  28  40   4  ...  19.0  3.0   
3  allisdo01    1871      1    WS3  NaN  27  133  28  44  10  ...  27.0  1.0   
4  ansonca01    1871      1    RC1  NaN  25  120  29  39  11  ...  16.0  6.0   

    CS  BB   SO  IBB  HBP  SH  SF  GIDP  
0  0.0   0  0.0  NaN  NaN NaN NaN   0.0  
1  1.0   4  0.0  NaN  NaN NaN NaN   0.0  
2  1.0   2  5.0  NaN  NaN NaN NaN   1.0  
3  1.0   0  2.0  NaN  NaN NaN NaN   0.0  
4  2.0   2  1.0  NaN  NaN NaN NaN   0.0  

[5 rows x 22 columns]

**Review the structure of batting. Pay close attention to how columns that start with a number get an 'X' in front of them! You'll need to know this to call those columns!**

In [4]:
# Rename columns starting with a number and place 'X' in front of them
batting.rename({'2B':'X2B', '3B':'X3B'}, inplace=True, axis=1)
# Output the structure of batting using .info()
batting.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105861 entries, 0 to 105860
Data columns (total 22 columns):
playerID    105861 non-null object
yearID      105861 non-null int64
stint       105861 non-null int64
teamID      105861 non-null object
lgID        105123 non-null object
G           105861 non-null int64
AB          105861 non-null int64
R           105861 non-null int64
H           105861 non-null int64
X2B         105861 non-null int64
X3B         105861 non-null int64
HR          105861 non-null int64
RBI         105105 non-null float64
SB          103493 non-null float64
CS          82320 non-null float64
BB          105861 non-null int64
SO          103761 non-null float64
IBB         69210 non-null float64
HBP         103044 non-null float64
SH          99792 non-null float64
SF          69757 non-null float64
GIDP        80420 non-null float64
dtypes: float64(9), int64(10), object(3)
memory usage: 17.8+ MB


**Call the head of the doubles (X2B) column**

In [5]:
# Return the first 5 rows of the X2B column
batting.X2B.head()

0     0
1     6
2     4
3    10
4    11
Name: X2B, dtype: int64

## Feature Engineering

We need to add three more statistics that were used in Moneyball! These are:

* [Batting Average](https://en.wikipedia.org/wiki/Batting_average)
* [On Base Percentage](https://en.wikipedia.org/wiki/On-base_percentage)
* [Slugging Percentage](https://en.wikipedia.org/wiki/Slugging_percentage)

Click on the links provided and search the wikipedia page for the formula for creating the new metric!

Which means that the Batting Average is equal to H (Hits) divided by AB (At Base).

**Create a new column with the Batting Average metric. Name this Column 'BA'.**

In [6]:
# BattingAverage if the amount of actual Hits over a player's At Bats (~Attempts)
batting['BA'] = batting['H']/batting['AB']

**After doing this operation, check the last 5 entries of the BA column of your data frame.**

In [7]:
# Using .tail() we retrieve the last five records
batting.BA.tail()

105856    0.226415
105857    0.000000
105858    0.263889
105859    0.305495
105860    0.201072
Name: BA, dtype: float64

Now do the same for some new columns! On Base Percentage (OBP) and Slugging Percentage (SLG). Hint: For SLG, you need 1B (Singles), this isn't in your data frame. However you can calculate it by subtracting doubles,triples, and home runs from total hits (H): 1B = H-2B-3B-HR

* **Create an OBP Column**
* **Create an SLG Column**

In [8]:
# On Base Percentage = (Hits+BasesOnBalls+HitByPitch) / (At Bats+BasesOnBalls+HitByPitch+SacrificeFlies)
batting['OBP'] = (batting.H + batting.BB + batting.HBP) / (batting.AB + batting.BB + batting.HBP + batting.SF)

In [9]:
# Creating X1B (Singles)
batting['X1B'] = batting['H'] - (batting['HR'] + batting['X3B'] + batting['X2B'])

In [10]:
# Creating Slugging Average (SLG)
batting['SLG'] =  ((4*batting['HR']) + (3*batting['X3B']) + (2*batting['X2B']) + (1*batting['X1B'])) / batting['AB']

**Check the structure of your data frame using .info()**

In [11]:
batting.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105861 entries, 0 to 105860
Data columns (total 26 columns):
playerID    105861 non-null object
yearID      105861 non-null int64
stint       105861 non-null int64
teamID      105861 non-null object
lgID        105123 non-null object
G           105861 non-null int64
AB          105861 non-null int64
R           105861 non-null int64
H           105861 non-null int64
X2B         105861 non-null int64
X3B         105861 non-null int64
HR          105861 non-null int64
RBI         105105 non-null float64
SB          103493 non-null float64
CS          82320 non-null float64
BB          105861 non-null int64
SO          103761 non-null float64
IBB         69210 non-null float64
HBP         103044 non-null float64
SH          99792 non-null float64
SF          69757 non-null float64
GIDP        80420 non-null float64
BA          89521 non-null float64
OBP         53929 non-null float64
X1B         105861 non-null int64
SLG         89521 non

## Merging Salary Data with Batting Data

We know we don't just want the best players, we want the most undervalued players, meaning we will also need to know current salary information! We have salary information in a csv file.

Complete the following steps to merge the salary data with the player stats!

**Load the Salaries.csv file into a dataframe called sal.**

In [12]:
sal_url = 'https://rotterdamai001.blob.core.windows.net/python/mlb/core/Salaries.csv'

In [13]:
sal = pd.read_csv(sal_url)

**Use describe to get a summary of the batting data frame and notice the minimum year in the yearID column. Our batting data goes back to 1871! Our salary data starts at 1985, meaning we need to remove the batting data that occured before 1985.**

**Reassign batting to only contain data from 1985 and onwards**

In [14]:
# Get the summary statistics on the dataset
batting.describe()

yearID          stint              G             AB  \
count  105861.000000  105861.000000  105861.000000  105861.000000   
mean     1965.775290       1.078773      51.214338     140.960694   
std        39.319486       0.286613      47.057599     184.433173   
min      1871.000000       1.000000       1.000000       0.000000   
25%      1935.000000       1.000000      13.000000       4.000000   
50%      1975.000000       1.000000      34.000000      48.000000   
75%      2000.000000       1.000000      80.000000     229.000000   
max      2018.000000       5.000000     165.000000     716.000000   

                   R              H            X2B            X3B  \
count  105861.000000  105861.000000  105861.000000  105861.000000   
mean       18.694212      36.861583       6.266321       1.271875   
std        28.180404      52.471260       9.666868       2.621256   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         4.000000       9.000000       1.000000       0.000000   
75%        27.000000      57.000000       9.000000       1.000000   
max       198.000000     262.000000      67.000000      36.000000   

                  HR            RBI  ...             SO           IBB  \
count  105861.000000  105105.000000  ...  103761.000000  69210.000000   
mean        2.843209      16.968432  ...      20.577057      1.086086   
std         6.354058      26.342342  ...      28.345666      2.750420   
min         0.000000       0.000000  ...       0.000000      0.000000   
25%         0.000000       0.000000  ...       1.000000      0.000000   
50%         0.000000       3.000000  ...       9.000000      0.000000   
75%         2.000000      24.000000  ...      29.000000      1.000000   
max        73.000000     191.000000  ...     223.000000    120.000000   

                 HBP            SH            SF        GIDP            BA  \
count  103044.000000  99792.000000  69757.000000  80420.0000  89521.000000   
mean        1.060809      2.252535      1.043594      2.9404      0.208322   
std         2.290011      4.198339      1.950372      4.7085      0.122852   
min         0.000000      0.000000      0.000000      0.0000      0.000000   
25%         0.000000      0.000000      0.000000      0.0000      0.145833   
50%         0.000000      0.000000      0.000000      0.0000      0.230769   
75%         1.000000      3.000000      1.000000      4.0000      0.273973   
max        51.000000     67.000000     19.000000     36.0000      1.000000   

                OBP            X1B           SLG  
count  53929.000000  105861.000000  89521.000000  
mean       0.259380      26.480177      0.290522  
std        0.145034      37.334650      0.188487  
min        0.000000       0.000000      0.000000  
25%        0.186441       0.000000      0.176471  
50%        0.290566       7.000000      0.309524  
75%        0.337079      42.000000      0.398773  
max        1.000000     225.000000      4.000000  

[8 rows x 23 columns]

In [15]:
# Take a subset of the batting dataframe and save it, again, into batting
batting = batting[batting.yearID >= 1985]

**Now use .describe() again to make sure the subset reassignment worked, your yearID min should be 1985.**

In [16]:
# Get the summary statistics on the dataset
batting.describe()

yearID         stint             G            AB             R  \
count  43606.000000  43606.000000  43606.000000  43606.000000  43606.000000   
mean    2002.634821      1.083291     50.634958    122.732629     16.381759   
std        9.660860      0.290565     46.010253    180.753142     26.928199   
min     1985.000000      1.000000      1.000000      0.000000      0.000000   
25%     1995.000000      1.000000     14.000000      0.000000      0.000000   
50%     2003.000000      1.000000     33.000000     21.000000      1.000000   
75%     2011.000000      1.000000     75.000000    186.000000     22.000000   
max     2018.000000      5.000000    163.000000    716.000000    152.000000   

                  H           X2B           X3B            HR           RBI  \
count  43606.000000  43606.000000  43606.000000  43606.000000  43606.000000   
mean      32.016787      6.177567      0.689676      3.595560     15.528689   
std       50.462246     10.204954      1.609161      7.401939     26.429059   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        3.000000      0.000000      0.000000      0.000000      1.000000   
75%       46.000000      9.000000      1.000000      3.000000     20.000000   
max      262.000000     59.000000     23.000000     73.000000    165.000000   

       ...            SO           IBB           HBP            SH  \
count  ...  43606.000000  43606.000000  43606.000000  43606.000000   
mean   ...     23.864973      0.949342      1.109801      1.172889   
std    ...     34.818562      2.672178      2.462070      2.397355   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...      0.000000      0.000000      0.000000      0.000000   
50%    ...      6.000000      0.000000      0.000000      0.000000   
75%    ...     35.000000      1.000000      1.000000      1.000000   
max    ...    223.000000    120.000000     35.000000     39.000000   

                 SF          GIDP            BA           OBP           X1B  \
count  43606.000000  43606.000000  31883.000000  31985.000000  43606.000000   
mean       1.010044      2.751021      0.202672      0.258944     21.553983   
std        1.940023      4.673531      0.131209      0.151523     33.837804   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.131579      0.181818      0.000000   
50%        0.000000      0.000000      0.230769      0.293774      2.000000   
75%        1.000000      4.000000      0.272727      0.339161     31.000000   
max       17.000000     35.000000      1.000000      1.000000    225.000000   

                SLG  
count  31883.000000  
mean       0.301263  
std        0.208998  
min        0.000000  
25%        0.160000  
50%        0.333333  
75%        0.422481  
max        4.000000  

[8 rows x 23 columns]

Now it is time to merge the batting data with the salary data! Since we have players playing multiple years, we'll have repetitions of playerIDs for multiple years, meaning we want to merge on both players and years.

**Use the pd.merge() function to merge the batting and sal data frames by ['playerID','yearID','teamID']. Call the new data frame combo.**

In [17]:
# We merge based on three columns, as a plaer may have played for 2 teams in one year
combo = batting.merge(sal,on=['playerID','yearID','teamID'])
combo.head()

playerID  yearID  stint teamID lgID_x   G   AB   R   H  X2B  ...  HBP  \
0  ackerji01    1985      1    TOR     AL  61    0   0   0    0  ...  0.0   
1  agostju01    1985      1    CHA     AL  54    0   0   0    0  ...  0.0   
2  aguaylu01    1985      1    PHI     NL  91  165  27  46    7  ...  6.0   
3  alexado01    1985      1    TOR     AL  36    0   0   0    0  ...  0.0   
4  allenne01    1985      1    SLN     NL  23    2   0   0    0  ...  0.0   

    SH   SF  GIDP        BA       OBP  X1B       SLG  lgID_y  salary  
0  0.0  0.0   0.0       NaN       NaN    0       NaN      AL  170000  
1  0.0  0.0   0.0       NaN       NaN    0       NaN      AL  147500  
2  4.0  3.0   7.0  0.278788  0.377551   30  0.466667      NL  237000  
3  0.0  0.0   0.0       NaN       NaN    0       NaN      AL  875000  
4  0.0  0.0   0.0  0.000000  0.000000    0  0.000000      NL  750000  

[5 rows x 28 columns]

**Retrieve the summary statistics of our new combo data.**

In [18]:
combo.describe()

yearID         stint             G            AB             R  \
count  25441.000000  25441.000000  25441.000000  25441.000000  25441.000000   
mean    2000.917535      1.007625     67.843717    173.077473     23.346567   
std        8.917739      0.094362     47.616122    205.066987     31.100341   
min     1985.000000      1.000000      1.000000      0.000000      0.000000   
25%     1994.000000      1.000000     29.000000      1.000000      0.000000   
50%     2001.000000      1.000000     56.000000     63.000000      5.000000   
75%     2009.000000      1.000000    107.000000    331.000000     41.000000   
max     2016.000000      4.000000    163.000000    716.000000    152.000000   

                  H           X2B           X3B            HR           RBI  \
count  25441.000000  25441.000000  25441.000000  25441.000000  25441.000000   
mean      45.643882      8.815888      0.971306      5.132896     22.280885   
std       57.922592     11.766046      1.897025      8.702250     30.694146   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%       11.000000      2.000000      0.000000      0.000000      5.000000   
75%       85.000000     16.000000      1.000000      7.000000     38.000000   
max      262.000000     59.000000     23.000000     73.000000    165.000000   

       ...           IBB           HBP            SH            SF  \
count  ...  25441.000000  25441.000000  25441.000000  25441.000000   
mean   ...      1.434102      1.554459      1.636846      1.462600   
std    ...      3.295388      2.884414      2.843783      2.279098   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...      0.000000      0.000000      0.000000      0.000000   
50%    ...      0.000000      0.000000      0.000000      0.000000   
75%    ...      1.000000      2.000000      2.000000      2.000000   
max    ...    120.000000     35.000000     39.000000     17.000000   

               GIDP            BA           OBP           X1B           SLG  \
count  25441.000000  19858.000000  19915.000000  25441.000000  19858.000000   
mean       3.915019      0.211102      0.269012     30.723792      0.315935   
std        5.408808      0.119831      0.140616     38.879314      0.191636   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.159091      0.206897      0.000000      0.200000   
50%        1.000000      0.241573      0.304734      9.000000      0.351974   
75%        7.000000      0.275641      0.344444     57.000000      0.431233   
max       35.000000      1.000000      1.000000    225.000000      4.000000   

             salary  
count  2.544100e+04  
mean   2.114363e+06  
std    3.476874e+06  
min    0.000000e+00  
25%    3.000000e+05  
50%    5.750000e+05  
75%    2.400000e+06  
max    3.300000e+07  

[8 rows x 24 columns]

## Analyzing lost players

The Oakland A's lost 3 key players during the off-season. We'll want to get their stats to see what we have to replace. The players lost were: first baseman 2000 AL MVP Jason Giambi (giambja01) to the New York Yankees, outfielder Johnny Damon (damonjo01) to the Boston Red Sox and infielder Rainer Gustavo "Ray" Olmedo ('saenzol01').

**Get a data frame called lost_players from the combo data frame consisting of those 3 players.**

In [19]:
lost_players_Ids = ['giambja01','damonjo01','saenzol01']
lost_players = combo[combo.playerID.isin(lost_players_Ids)]

**Review the first 40 rows of lost_players.**

In [20]:
# use .head() with a parameter for n of rows
lost_players.head(40)

playerID  yearID  stint teamID lgID_x    G   AB    R    H  X2B  ...  \
7295   damonjo01    1995      1    KCA     AL   47  188   32   53   11  ...   
7392   giambja01    1995      1    OAK     AL   54  176   27   45    7  ...   
8246   damonjo01    1996      1    KCA     AL  145  517   61  140   22  ...   
8330   giambja01    1996      1    OAK     AL  140  536   84  156   40  ...   
9156   damonjo01    1997      1    KCA     AL  146  472   70  130   12  ...   
9238   giambja01    1997      1    OAK     AL  142  519   66  152   41  ...   
10055  damonjo01    1998      1    KCA     AL  161  642  104  178   30  ...   
10146  giambja01    1998      1    OAK     AL  153  562   92  166   28  ...   
11020  damonjo01    1999      1    KCA     AL  145  583  101  179   39  ...   
11111  giambja01    1999      1    OAK     AL  158  575  115  181   36  ...   
11585  saenzol01    1999      1    OAK     AL   97  255   41   70   18  ...   
11948  damonjo01    2000      1    KCA     AL  159  655  136  214   42  ...   
12031  giambja01    2000      1    OAK     AL  152  510  108  170   29  ...   
12424  saenzol01    2000      1    OAK     AL   76  214   40   67   12  ...   
12749  damonjo01    2001      1    OAK     AL  155  644  108  165   34  ...   
12830  giambja01    2001      1    OAK     AL  154  520  109  178   47  ...   
13237  saenzol01    2001      1    OAK     AL  106  305   33   67   21  ...   
13560  damonjo01    2002      1    BOS     AL  154  623  118  178   34  ...   
13639  giambja01    2002      1    NYA     AL  155  560  120  176   34  ...   
14038  saenzol01    2002      1    OAK     AL   68  156   15   43   10  ...   
14365  damonjo01    2003      1    BOS     AL  145  608  103  166   32  ...   
14445  giambja01    2003      1    NYA     AL  156  535   97  134   25  ...   
15161  damonjo01    2004      1    BOS     AL  150  621  123  189   35  ...   
15254  giambja01    2004      1    NYA     AL   80  264   33   55    9  ...   
15975  damonjo01    2005      1    BOS     AL  148  624  117  197   35  ...   
16054  giambja01    2005      1    NYA     AL  139  417   74  113   14  ...   
16435  saenzol01    2005      1    LAN     NL  109  319   39   84   24  ...   
16767  damonjo01    2006      1    NYA     AL  149  593  115  169   35  ...   
16855  giambja01    2006      1    NYA     AL  139  446   92  113   25  ...   
17222  saenzol01    2006      1    LAN     NL  103  179   30   53   15  ...   
17556  damonjo01    2007      1    NYA     AL  141  533   93  144   27  ...   
17640  giambja01    2007      1    NYA     AL   83  254   31   60    8  ...   
18023  saenzol01    2007      1    LAN     NL   92  110    9   21    5  ...   
18372  damonjo01    2008      1    NYA     AL  143  555   95  168   27  ...   
18459  giambja01    2008      1    NYA     AL  145  458   68  113   19  ...   
19176  damonjo01    2009      1    NYA     AL  143  550  107  155   36  ...   
19258  giambja01    2009      1    OAK     AL   83  269   39   52   13  ...   
19961  damonjo01    2010      1    DET     AL  145  539   81  146   36  ...   
20042  giambja01    2010      1    COL     NL   87  176   17   43    9  ...   
20758  damonjo01    2011      1    TBA     AL  150  582   79  152   29  ...   

        HBP    SH    SF  GIDP        BA       OBP  X1B       SLG  lgID_y  \
7295    1.0   2.0   3.0   2.0  0.281915  0.323529   34  0.441489      AL   
7392    3.0   1.0   2.0   4.0  0.255682  0.363636   32  0.397727      AL   
8246    3.0  10.0   5.0   4.0  0.270793  0.312950  107  0.367505      AL   
8330    5.0   1.0   5.0  15.0  0.291045  0.355109   95  0.481343      AL   
9156    3.0   6.0   1.0   3.0  0.275424  0.337838  102  0.385593      AL   
9238    6.0   0.0   8.0  11.0  0.292871  0.362245   89  0.495183      AL   
10055   4.0   3.0   3.0   4.0  0.277259  0.339463  120  0.439252      AL   
10146   5.0   0.0   9.0  16.0  0.295374  0.383562  111  0.489324      AL   
11020   3.0   3.0   4.0  13.0  0.307033  0.378995  117  0.476844      AL   
11111   

Since all these players were lost in after 2001 in the offseason, let's only concern ourselves with the data from 2001.

**Save a subset of lost_players again to only grab the rows where the yearID was 2001.**

In [21]:
lost_players[lost_players.yearID == 2001]

playerID  yearID  stint teamID lgID_x    G   AB    R    H  X2B  ...  \
12749  damonjo01    2001      1    OAK     AL  155  644  108  165   34  ...   
12830  giambja01    2001      1    OAK     AL  154  520  109  178   47  ...   
13237  saenzol01    2001      1    OAK     AL  106  305   33   67   21  ...   

        HBP   SH   SF  GIDP        BA       OBP  X1B       SLG  lgID_y  \
12749   5.0  5.0  4.0   7.0  0.256211  0.323529  118  0.363354      AL   
12830  13.0  0.0  9.0  17.0  0.342308  0.476900   91  0.659615      AL   
13237  13.0  1.0  3.0   9.0  0.219672  0.291176   36  0.383607      AL   

        salary  
12749  7100000  
12830  4103333  
13237   290000  

[3 rows x 28 columns]

**Reduce the lost_players data frame to only include the following columns: playerID,H,X2B,X3B,HR,OBP,SLG,BA,AB**

In [22]:
# Using .loc[row_indexer,column_indexer] we reduce the DataFrame
include_columns = ['playerID','H','X2B','X3B','HR','OBP','SLG','BA','AB']
lost_players = lost_players.loc[:,include_columns]

**Print the .head() of lost_players dataframe**

In [23]:
lost_players.head()

playerID    H  X2B  X3B  HR       OBP       SLG        BA   AB
7295  damonjo01   53   11    5   3  0.323529  0.441489  0.281915  188
7392  giambja01   45    7    0   6  0.363636  0.397727  0.255682  176
8246  damonjo01  140   22    5   6  0.312950  0.367505  0.270793  517
8330  giambja01  156   40    1  20  0.355109  0.481343  0.291045  536
9156  damonjo01  130   12    8   8  0.337838  0.385593  0.275424  472

## Replacement Players

Now we have all the information we need! Here is your final task - Find Replacement Players for the key three players we lost! However, you have three constraints:

* **The total combined salary of the three players can not exceed 15 million dollars.**
* **Their combined number of At Bats (AB) needs to be equal to or greater than the lost players.**
* **Their mean OBP had to equal to or greater than the mean OBP of the lost players.**

Use the combo dataframe you previously created as the source of information! Remember to just use the 2001 subset of that dataframe. There's lost of different ways you can do this, so be creative! It should be relatively simple to find 3 players that satisfy the requirements, note that there are many correct combinations available!

There are a lot of correct answers for this part! This is where you can really have fun and explore the data with ggplot, figure out which are good data points to split your data on to find replacement players. This ending is left intentionally more open-ended so you can get a feel for exploring real data!

In [24]:
print("Average AB",lost_players.AB.mean())
print("Average OBP",lost_players.OBP.mean())

Average AB 424.4186046511628
Average OBP 0.36510441063193805


In [25]:
# Reduce the dataset to include statistics data of season of the year 2001
combo = combo[combo.yearID==2001]
# Let us focus on the most important columns for this task
combo = combo.loc[:,['playerID','AB','OBP','salary']]
# We remove the players we've lost from the DataFrame, NOTE: the use of ~ as a negator
combo = combo[~combo.playerID.isin(lost_players_Ids)]
# Retrieve players where AB >= than ta bit less than the average
combo = combo[combo.AB > lost_players.AB.mean()]
# I exclude players with a salary higher than 8 million
combo = combo[combo.salary <= 7000000]
# Retrieve data where OBP larger thant 0.34
combo = combo[combo.OBP >= lost_players.OBP.mean()]
# Create a new metric that combines both AB & OBP
combo['MAX'] = combo['AB'] * combo['OBP']
combo.sort_values(by='MAX', ascending=False)

playerID   AB       OBP   salary         MAX
13297  suzukic01  692  0.381471  5666667  263.978202
12847  gonzalu01  609  0.428571  4833333  261.000000
12894  heltoto01  587  0.431655  4950000  253.381295
12651  berkmla01  577  0.430233   305000  248.244186
12622  bagweje01  600  0.397490  6500000  238.493724
13181  pujolal01  590  0.402963   200000  237.748148
13288  stewash01  640  0.371060  2183333  237.478510
12619  aurilri01  636  0.368805  3250000  234.559767
13167  pierrju01  617  0.378176   215000  233.334828
12666  boonebr01  623  0.372263  3250000  231.919708
12593  abreubo01  588  0.393466  4983000  231.357955
13115  olerujo01  572  0.401180  6700000  229.474926
13122  ordonma01  593  0.381520  3750000  226.241431
12865  guerrvl01  599  0.377049  6000000  225.852459
12963   kentje01  607  0.369253  6000000  224.136494
12809  floydcl01  555  0.389507  5500000  216.176471
12838  glaustr01  588  0.367232  1250000  215.932203
13101  nevinph01  546  0.387821  1625000  211.750000
13066  mientdo01  543  0.386581   215000  209.913738
13226  rolensc01  554  0.378254  5500000  209.552833
13299  sweenmi01  559  0.374010  5000000  209.071315
12968  kleskry01  538  0.384013  5750000  206.598746
12782  edmonji01  500  0.410214  6333333  205.107084
12604   aloumo01  513  0.395869  5250000  203.080895
12727  coninje01  524  0.386023  3000000  202.276206
12857  grievbe01  542  0.372457  2750000  201.871674
13104  nixontr01  535  0.376396   390000  201.371611
13031  martied01  470  0.423408  5500000  199.001721
12909  higgibo02  541  0.367089  5325000  198.594937
12701  caseyse01  533  0.369048  3000000  196.702381
12851  gracema01  476  0.385870  3000000  183.673913
12708  catalfr01  463  0.391389   850000  181.213307
13339  vidrojo01  486  0.370510  2500000  180.068053
13060  menecfr01  471  0.368696   225000  173.655652
13003  loducpa01  460  0.373541   230000  171.828794
13067  millake01  449  0.373737   700000  167.808081
12688  burksel01  439  0.368932  5666667  161.961165
12718  clarkto02  428  0.374245  4462500  160.177062

In [26]:
# From this list I choose berkmla01, pujolal01, pierrju01. In this way we may throw a big party with the leftover budget.
selected_players = ['berkmla01','pujolal01','pierrju01']
combo = combo[combo.playerID.isin(selected_players)]
combo.describe()

AB       OBP         salary         MAX
count    3.000000  3.000000       3.000000    3.000000
mean   594.666667  0.403791  240000.000000  239.775721
std     20.404248  0.026038   56789.083458    7.658690
min    577.000000  0.378176  200000.000000  233.334828
25%    583.500000  0.390570  207500.000000  235.541488
50%    590.000000  0.402963  215000.000000  237.748148
75%    603.500000  0.416598  260000.000000  242.996167
max    617.000000  0.430233  305000.000000  248.244186

In [27]:
# About that party...
budget = 15000000
print('The remaining budget for a party is', budget - combo.salary.sum())

The remaining budget for a party is 14280000


## More exercises like these?

This project was inspired by Jose Portilla's Data Science and Machine learning in R course un [Udemy](https://www.udemy.com/course/data-science-and-machine-learning-bootcamp-with-r/learn/lecture/5412744?start=180#overview). Jose has great content regarding Data Science using R as well as Python. If you liked this assignment and want to be challenged more please find one of his courses through the Udemy platform.

![ImageOfUdemy](https://miro.medium.com/max/1200/1*HuQyl7_WMMzOfs8RIlQ-XA.png)